src/Adapter.php

```diff
@@ -91,6 +91,44 @@ public function removePolicy($sec, $ptype, $rule): void
         $this->table->delete($entity);
     }
 
+    /**
+     * @param string $sec
+     * @param string $ptype
+     * @param int $fieldIndex
+     * @param string|null ...$fieldValues
+     * @return array
+     * @throws Throwable
+     */
+    public function _removeFilteredPolicy(string $sec, string $ptype, int $fieldIndex, ?string ...$fieldValues): array
+    {
+        $removedRules = [];
+        $this->table->getConnection()->transactional(function () use ($sec, $ptype, $fieldIndex, $fieldValues, &$removedRules) {
+            $entity = $this->table->find()->where(['ptype' => $ptype]);
+
+            foreach ($fieldValues as $value) {
+                if (!is_null($value) && $value !== '') {
+                    $entity->where(['v' . strval($fieldIndex) => $value]);
+                }
+                $fieldIndex++;
+            }
+
+            $removedRules = [];
+            $rows = $entity->all();
+            foreach ($rows as $row) {
+                unset($row->id);
+                unset($row->ptype);
+                $row = $row->toArray();
+                $row = array_filter($row, function ($value) {
+                    return !is_null($value) && $value !== '';
+                });
+                $removedRules[] = array_values($row);
+            }
+            $this->removePolicies($sec, $ptype, $removedRules);
```

Comment:
Here again use `removePolicies` to delete one by one, which will affect performance .